packages/driver/src/cy/commands/actions/attachFile.ts

```diff
@@ -6,23 +6,100 @@ import $errUtils from '../../../cypress/error_utils'
+/* dropzone.js relies on an experimental, nonstandard API, webkitGetAsEntry().
+ * https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/webkitGetAsEntry
+ * The behavior here varies between browsers, and though unpleasant, we have to attempt
+ * to replicate it as far as possible.
+ *
+ * In Firefox, attempting to set webkitGetAsEntry() fails - but this is fine, because
+ * Firefox also returns a useful value from item.webkitGetAsEntry().
+ *
+ * In webkit browsers however, item.webkitGetAsEntry() returns null, because our File()
+ * instances don't actually correspond with a file on disk. But the property is writeable,
+ * so we're able to override it with our own implementation that returns a proper object.
+ */
+const tryMockWebkit = (item) => {
+  try {
+    item.webkitGetAsEntry = () => {
+      return {
+        isFile: true,
+        file: (callback) => callback(item.getAsFile()),
+      }
+    }
+  } catch (e) {
+    // We're in Firefox, this is fine.
+  }
+
+  return item
+}
+
+const createDataTransfer = (files: Cypress.FileReferenceObject[]): DataTransfer => {
   const dataTransfer = new DataTransfer()
 
   files.forEach(({ contents, fileName, lastModified, mimeType }) => {
-    const file = new File([contents], fileName || '', { lastModified, type: mimeType })
+    dataTransfer.items.add(new File([contents], fileName || '', {
+      lastModified,
+      type: mimeType,
+    }))
+  })
+
+  const oldItems = dataTransfer.items
 
-    dataTransfer.items.add(file)
+  // dataTransfer.items is a getter, and it - and the items read from its
+  // returned DataTransferItemList - cannot be assigned to. DataTransferItemLists
+  // also cannot be constructed, so we have to use an array instead.
+  Object.defineProperty(dataTransfer, 'items', {
+    get () {
+      return _.map(oldItems, tryMockWebkit)
+    },
   })
 
-  return dataTransfer.files
+  return dataTransfer
 }
 
 interface InternalAttachFileOptions extends Cypress.AttachFileOptions {
   _log: any
   $el: JQuery
 }
 
+const ACTIONS = {
+  input: (element, dataTransfer, coords, state) => {
+    (element as HTMLInputElement).files = dataTransfer.files
+    const inputEventOptions = addEventCoords({
+      bubbles: true,
+      composed: true,
+    }, coords)
+
+    const changeEventOptions = addEventCoords({
+      bubbles: true,
+    }, coords)
+
+    dispatch(element, state('window'), 'input', inputEventOptions)
+    dispatch(element, state('window'), 'change', changeEventOptions)
+  },
+  'drag-n-drop': (element, dataTransfer, coords, state) => {
+    const dragEventOptions = addEventCoords({
+      bubbles: true,
+      composed: true,
+      cancelable: true,
+      dataTransfer,
+    }, coords)
+
+    dispatch(element, state('window'), 'drag', dragEventOptions)
+    dispatch(element, state('window'), 'dragenter', dragEventOptions)
+    dispatch(element, state('window'), 'dragover', dragEventOptions)
+    dispatch(element, state('window'), 'drop', dragEventOptions)
```

Comment:
I actually came across this project yesterday that we recommend in the docs, https://github.com/abramenal/cypress-file-upload, the drag & drop implementation here also [triggers dragLeave, dragEnd](https://github.com/abramenal/cypress-file-upload/blob/main/src/constants.ts#L17). Is this something we should be doing as well?

Reply:
It also looks like there is/was an issue with firefox not recieving the events without an additional `change` event: https://github.com/abramenal/cypress-file-upload/blob/48d5bdd0e4215345224faa490f5eedfc899533a3/src/helpers/attachFileToElement.ts#L15

Reply:
Also just lookin at this project, we'll likely need to log an issue to update their commands to be overrides to work correctly once this is released: https://github.com/abramenal/cypress-file-upload/blob/48d5bdd0e4215345224faa490f5eedfc899533a3/src/index.ts#L5

Reply:
> I actually came across this project yesterday that we recommend in the FAQ docs, abramenal/cypress-file-upload, the drag & drop implementation here also triggers dragLeave, dragEnd. Is this something we should be doing as well?

In my testing, these events to *not* trigger when dropping a file in a browser manually. Left t.hem out intentionally

> It also looks like there is/was an issue with firefox not recieving the events without an additional change event: abramenal/cypress-file-upload@48d5bdd/src/helpers/attachFileToElement.ts#L15

Interesting, and very good catch. I don't think the solution cypress-upload-file adopted is the correct one though. I've added a different approach and a test to cover this case.

> Also just lookin at this project, we'll likely need to log an issue to update their commands to be overrides to work correctly once this is released: abramenal/cypress-file-upload@48d5bdd/src/index.ts#L5

Current plan is that he's going to deprecate the package. https://github.com/abramenal/cypress-file-upload/issues/338#issuecomment-976764093.